ts/src/base/Exchange.ts

```diff
@@ -4541,6 +4541,206 @@ export default class Exchange {
+                    const responseLength = response.length;
+                    if (this.verbose) {
+                        this.log ('Dynamic pagination call', calls, 'method', method, 'response length', responseLength, 'timestamp', paginationTimestamp);
+                    }
+                    if (responseLength === 0) {
+                        break;
+                    }
+                    errors = 0;
+                    result = this.arrayConcat (result, response);
+                    const firstElement = this.safeValue (response, 0);
+                    paginationTimestamp = this.safeInteger2 (firstElement, 'timestamp', 0);
+                    if ((since !== undefined) && (paginationTimestamp <= since)) {
+                        break;
+                    }
+                } else {
+                    // do it forwards, starting from the since
+                    const response = await this[method] (symbol, paginationTimestamp, maxEntriesPerRequest, params);
+                    const responseLength = response.length;
+                    if (this.verbose) {
+                        this.log ('Dynamic pagination call', calls, 'method', method, 'response length', responseLength, 'timestamp', paginationTimestamp);
+                    }
+                    if (responseLength === 0) {
+                        break;
+                    }
+                    errors = 0;
+                    result = this.arrayConcat (result, response);
+                    const last = this.safeValue (response, responseLength - 1);
+                    paginationTimestamp = this.safeInteger (last, 'timestamp') - 1;
+                }
+            } catch (e) {
+                errors += 1;
+                if (errors > maxRetries) {
+                    throw e;
+                }
+            }
+        }
+        return this.removeRepeatedElementsFromArray (result);
+    }
+
+    async safeDeterministicCall (method, symbol = undefined, since = undefined, limit = undefined, timeframe = undefined, params = {}) {
+        let maxRetries = undefined;
+        [ maxRetries, params ] = this.handleOptionAndParams (params, method, 'maxRetries', 3);
+        let errors = 0;
+        try {
+            if (timeframe && method !== 'fetchFundingRateHistory') {
+                return await this[method] (symbol, timeframe, since, limit, params);
+            } else {
+                return await this[method] (symbol, since, limit, params);
+            }
+        } catch (e) {
+            errors += 1;
+            if (errors > maxRetries) {
+                throw e;
+            }
+        }
+    }
+
+    async fetchPaginatedCallDeterministic (method: string, symbol: string = undefined, since = undefined, limit = undefined, timeframe = undefined, params = {}, maxEntriesPerRequest = undefined): Promise<any> {
+        let maxCalls = undefined;
+        [ maxCalls, params ] = this.handleOptionAndParams (params, method, 'paginationCalls', 10);
```

Comment:
@pcriadoperez I think we might over-complicate by doing that,  let's say the `maxCalls` ends up being 500 or 1000 that would take forever and it would seem the program is hanged.

Reply:
Yeah thats true.. What about then notifying the user that if they fetched a since and until with not enough maxCalls that they won't receive all the data?

Reply:
Yes that I think we can do

Reply:
done